core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala

```diff
@@ -55,6 +60,9 @@ private[spark] class ApplicationPage(parent: MasterWebUI) extends WebUIPage("app
     val app = state.activeApps.find(_.id == appId).getOrElse({
       state.completedApps.find(_.id == appId).getOrElse(null)
     })
+    if (app == null) {
+      return Seq[Node]()
```

Comment:
Yeah that makes sense. I emulated the message in `HistoryServer`. Really this should result in an HTTP 404, but this simplified API doesn't give access to the `HttpServletResponse`. It wouldn't be a lot of work to pass this around all through the API. Worth it here, you guys think?